Dashboard: don't show the error toast when Jetpack site stats module is turned off

jaclync opened the pull request:
Fixes #932

## Changes
- Identified Jetpack site stats module disabled error at Yosemite layer (`SiteVisitStatsStoreError`) and to the app
- Handled `SiteVisitStatsStoreError` in `DashboardViewController`: in case of `statsModuleDisabled`, do nothing the first part for https://github.com/woocommerce/woocommerce-ios/issues/1070

## Testing
- Prerequisite: have a store with Jetpack installed
- Go to Jetpack modules settings page at your store's `/wp-admin/admin.php?page=jetpack_modules`, and find the "site stats" module
- Deactivate the "site stats" module
- Launch the app with the store --> there should not be a generic alert as in the screenshot [here](https://github.com/woocommerce/woocommerce-ios/issues/932#issuecomment-488332137)
- Go back to the Jetpack modules settings, and reactivate the "site stats" module --> the Dashboard should display the latest visitor stats

## Example screenshot

On the Dashboard (notice the toast at the bottom):

before | after (no toast)
--- | ---
![Simulator Screen Shot - iPhone SE - 2019-07-15 at 16 25 55](https://user-images.githubusercontent.com/1945542/61246633-51d03300-a71d-11e9-804d-60a4ff67ad97.png) | ![Simulator Screen Shot - iPhone SE - 2019-07-15 at 16 25 13](https://user-images.githubusercontent.com/1945542/61246636-5399f680-a71d-11e9-892a-7c52ac5cf9cb.png)


Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

ctarda commented on Networking/Networking/Remote/SiteVisitStatsRemote.swift:
```diff
@@ -48,3 +53,37 @@ private extension SiteVisitStatsRemote {
+        static let statsModuleDisabled: String = "This blog does not have the Stats module enabled"
+    }
+
+    init(error: Error) {
```
My first inclination would be making this method static, as in a [static factory method](https://stackify.com/static-factory-methods/). But, to be completely honest, I am not sure that would actually bring too many benefits.

jaclync replied:
I just read the example in the [static factory method](https://stackify.com/static-factory-methods/) post, and it seems like the benefits of the static factory method over initializer are:
- static factory method is a good way to provide different constructors for individual use cases, some with more arguments some with fewer --> in Swift, we have designated and convenience initializers
- it looks like the constructor in the example `TimeSpan` doesn't have a signature for each input argument, my guess is that the language doesn't support it. The static factory method provides a way to describe the argument(s) in the function signature --> in Swift, we can specify what each argument is in the signature with optional different external/internal naming

this is my interpretation of the post, I think Swift initializers address the issues in the example. Lemme know if I missed anything from the post / other benefits of static factory method!

ctarda replied:
I think that's pretty much it. As you mentioned, in this case I am not sure it actually brings much benefit.

ctarda commented on Networking/Networking/Remote/SiteVisitStatsRemote.swift:
```diff
@@ -48,3 +53,37 @@ private extension SiteVisitStatsRemote {
+        }
+        switch dotcomError {
+        case .unknown(let code, let message):
+            if code == ErrorIdentifiers.invalidBlog && message == ErrorMessages.statsModuleDisabled {
```
This seems prone to break, if the error message changes. I am not sure if it's even possible, or if it would be performant enough, or if it would even make sense, but what about relying on the Natural Language Framework to try to identify the likelihood of the error being the module disabled?

ctarda replied:
Oops, I just saw this other comment: https://github.com/woocommerce/woocommerce-ios/issues/1070#issuecomment-511566383

Please ignore me!

jaclync replied:
yes I'd much prefer if the check does not rely on a natural language message comparison. I hope in the future, the error code from the API is unique for each scenario so that the client side can handle each case in a less error-prone way

ctarda commented on Networking/Networking/Remote/SiteVisitStatsRemote.swift:
```diff
@@ -48,3 +53,37 @@ private extension SiteVisitStatsRemote {
+        }
+        switch dotcomError {
+        case .unknown(let code, let message):
+            if code == ErrorIdentifiers.invalidBlog && message == ErrorMessages.statsModuleDisabled {
```
I would probably pass up the error as it is, without trying to figure out if it is due to the plugin being disabled, and do this very same check in Yosemite. On the one hand, it seems to me like deciding the type of error that we receive, and what that error means to our system, is by definition business logic, and the networking layer should only move data to and from the backend, but on the other hand, maybe I'm overthinking it, what do you think?

mindgraffiti replied:
We already have a [little bit of business logic](https://github.com/woocommerce/woocommerce-ios/blob/develop/Networking/Networking/Remote/Remote.swift#L71-L75) in the main `Remote` class, so I would say there is precedent for this. You are our architect, so if you have your own preferences, let's follow that :)

ctarda replied:
Fair enough!

jaclync replied:
My reason for classifying the error at the Networking layer was that the error definition is from the API and translating the API error into what the API doc specifies felt like a networking concern. On the other hand, I see that this is business logic and other remote services don't seem to handle the errors at Networking layer. I'll move this to Yosemite layer!

jaclync replied:
Moved the error handling in this commit: https://github.com/woocommerce/woocommerce-ios/pull/1105/commits/a62110d618728519536ca6ce396df2995cce92f5

ctarda replied:
Thanks for taking the time to take the feedback into consideration!

ctarda commented on WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewController.swift:
```diff
@@ -271,6 +271,24 @@ private extension DashboardViewController {
         })
     }
 
+    private func handleSyncError(error: Error) {
```
This looks great. The UI handles an error that is generated by the only layer it is aware of: the business logic

jaclync commented:
Great feedback @ctarda and @mindgraffiti ! Just updated the PR:
- Merged `develop` to branch for merge conflicts in release notes
- Added unit tests in https://github.com/woocommerce/woocommerce-ios/pull/1105/commits/bf7ddf4b9c3c002d76fed910ba4a5a5a645f4b80
- Moved the error handling from Networking to Yosemite in https://github.com/woocommerce/woocommerce-ios/pull/1105/commits/a62110d618728519536ca6ce396df2995cce92f5 - I do like keeping Networking layer just making network requests without any business logic

And I just tested again. Ready for another pass!

ctarda commented:
I just run a quick test after the last update:
1. Deactivated site stats
2. There is no user facing error, neither when switching stores, nor when pulling down to refresh the dashboard, nor when killing and starting the app again.
3. Activated site stats again.
4. The visitors count is available again after a pull to refresh

:shipit: !
